packages/flutter/lib/src/material/menu_bar.dart

```diff
@@ -0,0 +1,3256 @@
+    if (focusedItem == null) {
+      return false;
+    }
+    if (focusedItem.parent == controller.root) {
+      // Pressing on a top level menu closes all the menus.
+      controller.openMenu = null;
+      return true;
+    }
+    _MenuNode? previousFocusable = focusedItem.previousSibling;
+    while (previousFocusable != null && !previousFocusable.focusNode!.canRequestFocus) {
+      previousFocusable = previousFocusable.previousSibling;
+    }
+    if (previousFocusable != null) {
+      controller.openMenu = previousFocusable;
+    } else if (focusedItem.parent?.parent == controller.root) {
+      // Pressing on a next-to-top level menu, moves to the parent.
+      controller.openMenu = focusedItem.parent;
+    }
+    return true;
+  }
+
+  bool _moveDown() {
+    final _MenuNode? focusedItem = controller.focusedItem;
+    if (focusedItem == null) {
+      return false;
+    }
+    if (focusedItem.parent == controller.root) {
+      if (controller.openMenu == null) {
+        controller.openMenu = focusedItem;
+        return true;
+      }
+      final List<_MenuNode> children = focusedItem.focusableChildren;
+      if (children.isNotEmpty) {
+        controller.openMenu = children[0];
+      }
+      return true;
+    }
+    _MenuNode? nextFocusable = focusedItem.nextSibling;
+    while (nextFocusable != null && !nextFocusable.focusNode!.canRequestFocus) {
+      nextFocusable = nextFocusable.nextSibling;
+    }
+    if (nextFocusable != null) {
+      controller.openMenu = nextFocusable;
+    }
+    return true;
+  }
+
+  @override
+  void invoke(DirectionalFocusIntent intent) {
+    final TextDirection textDirection = Directionality.of(controller.menuBarContext);
+    switch (intent.direction) {
+      case TraversalDirection.up:
+        if (_moveUp()) {
+          return;
+        }
+        break;
+      case TraversalDirection.down:
+        if (_moveDown()) {
+          return;
+        }
+        break;
+      case TraversalDirection.left:
+        switch (textDirection) {
+          case TextDirection.rtl:
+            if (_moveForward()) {
+              return;
+            }
+            break;
+          case TextDirection.ltr:
+            if (_moveBackward()) {
+              return;
+            }
+            break;
+        }
+        break;
+      case TraversalDirection.right:
+        switch (textDirection) {
+          case TextDirection.rtl:
+            if (_moveBackward()) {
+              return;
+            }
+            break;
+          case TextDirection.ltr:
+            if (_moveForward()) {
+              return;
+            }
+            break;
+        }
+
+        break;
+    }
+    super.invoke(intent);
+  }
+}
+
+/// Provides default implementation for [MenuItem] in a [StatefulWidget], to serve
+/// as a base class for the [MenuBarItem], [MenuItemGroup] and [MenuBarMenu] classes.
+abstract class _MenuBarItemDefaults extends StatefulWidget implements PlatformMenuItem {
```

Comment:
Maybe Base instead of Defaults to just indicate that it is the base class?

Reply:
I didn't think it could be a mixin because it needed a constructor, but actually it can if I make the members be getters instead. Done!